src/aes_gcm_plug.c

```diff
@@ -0,0 +1,414 @@
+/*
+ * Galois/Counter Mode (GCM) and GMAC with AES
+ *
+ * Copyright (c) 2012, Jouni Malinen <j@w1.fi>
+ *
+ * This software may be distributed under the terms of the BSD license.
+ *
+ * See https://w1.fi/hostapd/ for more information.
+ *
+ * For self-test use the following command,
+ * gcc -DTEST -I/usr/local/opt/openssl/include -L/usr/local/opt/openssl/lib aes_gcm_plug.c -lcrypt
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <openssl/aes.h>
+
+// #include "aes.h"
+
+#ifndef BIT
+#define BIT(x) (1U << (x))
+#endif
+
+
+static inline uint32_t WPA_GET_BE32(const uint8_t *a)
+{
+        return ((uint32_t) a[0] << 24) | (a[1] << 16) | (a[2] << 8) | a[3];
+}
+
+
+static inline void WPA_PUT_BE32(uint8_t *a, uint32_t val)
+{
+        a[0] = (val >> 24) & 0xff;
+        a[1] = (val >> 16) & 0xff;
+        a[2] = (val >> 8) & 0xff;
+        a[3] = val & 0xff;
+}
+
+
+static inline void WPA_PUT_BE64(uint8_t *a, uint64_t val)
+{
+        a[0] = val >> 56;
+        a[1] = val >> 48;
+        a[2] = val >> 40;
+        a[3] = val >> 32;
+        a[4] = val >> 24;
+        a[5] = val >> 16;
+        a[6] = val >> 8;
+        a[7] = val & 0xff;
+}
+
+
+static void inc32(uint8_t *block)
+{
+	uint32_t val;
+	val = WPA_GET_BE32(block + AES_BLOCK_SIZE - 4);
+	val++;
+	WPA_PUT_BE32(block + AES_BLOCK_SIZE - 4, val);
+}
+
+
+static void xor_block(uint8_t *dst, const uint8_t *src)
+{
+	uint32_t *d = (uint32_t *) dst;
+	uint32_t *s = (uint32_t *) src;
```

Comment:
These casts aren't alignment-safe. It will crash and burn on systems that requires alignment (unless all char buffers happen to be aligned, which seems to be the case with the format as well as the test code in this very file).

Reply:
Easy fix though: Something like this
```c
static void xor_block(uint8_t *dst, const uint8_t *src)
{
#if ARCH_ALLOWS_UNALIGNED
	uint32_t *d = (uint32_t *) dst;
	uint32_t *s = (uint32_t *) src;
	*d++ ^= *s++;
	*d++ ^= *s++;
	*d++ ^= *s++;
	*d++ ^= *s++;
#else
	int i;

	for (i = 0; i < 16; i++)
		*dst++ ^= *src++;
#endif
}
```

Reply:
Fix applied. Thanks!

I only have a AMD64 system at hand now.